### PR TITLE
revamp unit testing by running sodalint tests

### PR DESCRIPTION
**Pull request by kennywlo**
Return Full urls for SODA endpoints via flask's url_for

### CODE REVIEW

File: tests/sodalint/LICENSE [hunk 1]
```diff
@@ -0,0 +1,64 @@
+CSIRO Open Source Software Licence v1.0
+
```
Comment (cbanek) [inline tests/sodalint/LICENSE]: Why are you checking sodalint in?  Why not just download it for the test or as part of the setup, or put it in the container?  By checking in big files like this to the repo, it can slow you down as they never really get deleted.

Comment (cbanek) [reply tests/sodalint/LICENSE]: This comment applies to the .jar file mostly, but you also wouldn't have to check this in.

File: tests/test_soda_imgserv.py [hunk 2]
```diff
@@ -19,21 +19,14 @@
...
+    ret = False
+    assert(ret is True)
 
-    TODO: implement these tests for generic SODA
-    """
 
-    def test_1(self):
-        return
+def test_soda_2():
+    ret = False
```
Comment (cbanek) [inline tests/test_soda_imgserv.py]: So these tests still don't really exist, but now they fail instead of pass?  Does this get run as part of the CI?

**Comment by cbanek** [comment]
I believe I've mentioned this before, but please have multiple commits.  There is a lot of different things happening here in one commit.  Please separate out naming things if you can or deleting unused code.

File: python/lsst/dax/imgserv/api_soda.py [hunk 3]
```diff
@@ -88,67 +88,81 @@ def load_imgserv_config(config_path=None, metaserv_url=None):
...
 
 @image_soda.route("/")
-def index():
+def is_index():
     return make_response(render_template("api_image_soda.html"))
 
 
 @image_soda.route("/availability", methods=["GET"])
-def imagesoda_availability():
+def is_availability():
```
Comment (cbanek) [inline python/lsst/dax/imgserv/api_soda.py]: Why are you changing the names from imagesoda_ to is_?  Usually when I see a function that starts with is, it means it is a function that return a boolean.  Like, is_true, is_connected, etc.  I don't think that's what you're doing here?  I think getting rid of the is_ part and just call it availability, capabilities, etc, would be better

**Comment by timj** [comment]
See https://developer.lsst.io/work/flow.html#commits-should-represent-discrete-logical-changes-to-the-code for policy on commit organization.

File: python/lsst/dax/imgserv/api_soda.py [hunk 4]
```diff
@@ -161,19 +175,33 @@ def imagesoda_unhandled_exceptions(error):
     return resp
 
 
+def _service_response(soda_method):
+    """ Return service info using DALI template.
+    """
+    service_ep = os.getenv("IMGSERV_FQDN", "localhost:5000")
```
Comment (cbanek) [inline python/lsst/dax/imgserv/api_soda.py]: I thought we talked about using the Host header or you mentioned that flask had some special function to get the URL that you were called in on?

Comment (cbanek) [reply python/lsst/dax/imgserv/api_soda.py]: It really is worth it to not have to configure this if at all possible, since it's always different everywhere, and the container will never know it.

File: python/lsst/dax/imgserv/api_soda.py [hunk 5]
```diff
@@ -161,19 +175,33 @@ def imagesoda_unhandled_exceptions(error):
...
+    resp = make_response(render_template("dali_response.xml",
+                                         dali_resp_state="Error",
+                                         dali_resp_msg=message),
                          HTTPStatus.NOT_FOUND)
     resp.headers["Content-Type"] = "text/xml"
     return resp
 
 
-def _getparams():
+def _getparams(mark_api=True):
```
Comment (cbanek) [inline python/lsst/dax/imgserv/api_soda.py]: What is mark_api?  This seems very confusing?  Why did you add this as an option?  I don't see anyone passing in false

File: python/lsst/dax/imgserv/api_soda.py [hunk 6]
```diff
@@ -88,67 +88,81 @@ def load_imgserv_config(config_path=None, metaserv_url=None):
...
 
 
-@image_soda.route("/async", methods=["POST"])
-def imagesoda_async():
-    params = _getparams()
-    xml = current_app.soda.do_async(params)
+@image_soda.route("/async", methods=["GET", "POST"])
+def is_async():
+    if not request.args:
+        # no parameters present, return service status
```
Comment (cbanek) [inline python/lsst/dax/imgserv/api_soda.py]: Don't do this.  Just throw an error.  This is really not good because you aren't doing what you should be doing, but you'll return a 200?

File: python/lsst/dax/imgserv/api_soda.py [hunk 7]
```diff
@@ -88,67 +88,81 @@ def load_imgserv_config(config_path=None, metaserv_url=None):
...
+def is_sia():
+    resp = current_app.soda.do_sia(_getparams())
+    return make_response(resp)
+
+
 @image_soda.route("/sync", methods=["GET", "PUT", "POST"])
-def imagesoda_sync():
+def is_sync():
+    if not request.args:
+        # no parameters present, return service status
```
Comment (cbanek) [inline python/lsst/dax/imgserv/api_soda.py]: Same here, throw an error

File: python/lsst/dax/imgserv/templates/dali_examples.xml [hunk 8]
```diff
@@ -1,13 +1,9 @@
...
-      href="{{ url_for('api_image_soda.static', filename='style.css') }}">
-<div vocab="http://www.ivoa.net/rdf/examples#">
+<?xml version="1.0" encoding="UTF-8"?>
+<div vocab="http://www.ivoa.net/rdf/examples">
 
     <div id="brect" resource="#brect" typeof="example">
-    <p> Example 1: <h2 property="name">Synchronous Query by BRECT </h2>
-        <a href="{{ url_dali_sync }}">
+    <p> Example 1: <h3 property="name">Synchronous Query by BRECT </h3>
+        <a href="{{ url_for('api_image_soda.is_sync', _external=True) }}">
```
Comment (cbanek) [inline python/lsst/dax/imgserv/templates/dali_examples.xml]: Ah okay this looks like the flask part that does what you want.  This is good.  Although it would be less confusing if you got rid of the is_.  Does this mean the other thing to determine the capabilities is only called when someone doesn't pass in args?  How are these paths related to each other?

File: python/lsst/dax/imgserv/templates/vosi_capabilities.xml [hunk 9]
```diff
@@ -1,60 +1,64 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<?xml-stylesheet type="text/css" href= "{{ url_for('api_image_soda.static', filename='style.css') }}"?>
 <vosi:capabilities
    xmlns:vosi="http://www.ivoa.net/xml/VOSICapabilities/v1.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-   xmlns:vod="http://www.ivoa.net/xml/VODataService/v1.1">
+   xmlns:vod="http://www.ivoa.net/xml/VODataService/v1.1"
+   xmlns:sia="http://www.ivoa.net/xml/SIA/v1.1">
```
Comment (cbanek) [inline python/lsst/dax/imgserv/templates/vosi_capabilities.xml]: So you're adding in SIA?

File: python/lsst/dax/imgserv/vo/imageSODA.py [hunk 10]
```diff
@@ -85,6 +84,21 @@ def do_async(self, params: dict) -> object:
...
+        ----------
+        params : `dict`
+            the HTTP parameters.
+
+        Returns
+        -------
+        xml: `str`
+
+        """
+        raise NotImplementedError("ImageSODA.do_sia()")
```
Comment (cbanek) [inline python/lsst/dax/imgserv/vo/imageSODA.py]: So you're adding in SIA?  Or is this future work that will be done soon?

File: tests/conftest.py [hunk 11]
```diff
@@ -18,21 +18,25 @@
...
 
-class DALUnitTest(unittest.TestCase):
-    """Unit tests for DAL service components.
 
-    TODO: Implement these tests for generic SODA.
-    """
+@pytest.fixture
+def app():
+    app = Flask(__name__)
+    app.register_blueprint(is_api_soda.image_soda, url_prefix='/api/image/soda')
```
Comment (cbanek) [inline tests/conftest.py]: The URL that this is running at will likely need to be configurable in the future, but I guess this is fine for now

File: tests/engineFactory.py [hunk 12]
```diff
@@ -1,140 +0,0 @@
...
-    """
-    Initializes and returns SQLAlchemy engine using provided values.
-
-    To specify socket, use query={"unix_socket": "/the/socket.file"}
-
-    Note, if mysql sees "localhost" it switches to using socket, even if port is
-    specified. Commonly used way around it is to specify "127.0.0.1" as port for
-    local access.
-    """
-    url = URL(drivername=drivername,
```
Comment (cbanek) [inline tests/engineFactory.py]: What are some of these files that are deleted?  Is this some other unrelated cleanup?

Comment (cbanek) [reply tests/engineFactory.py]: Are you deleting all the tests because now we're using sodalint?  Does that cover all of this?

File: tests/sodalint/LICENSE (hunk 1 above)
Comment (kennywlo) [reply tests/sodalint/LICENSE]: The .jar file is 7MB, which is argubably no longer considered big today (LICENSE file added just in case).  sodalint is lightly used out there; other than GitHub, there is no public repo for its release binaries.  I'm aware of 'git checkout tags/thetagoftherelease', which can apply to sodalint here.  But once I learn how that's handled exactly in our Jenkins-CI env, it's something to be considered.

Comment (timj) [reply tests/sodalint/LICENSE]: At the very least, adding third party code to a repo should be a separate commit, such that it is easy to revert later on if you change the way you do the sodalint testing.

File: tests/test_soda_imgserv.py (hunk 2 above)
Comment (kennywlo) [reply tests/test_soda_imgserv.py]: Good catch - the intention is to update these tests once SODA operations ready for unit testing, which among other things, require dummy (small) FITS files for testing purpose. I've changed the code to reflect this.

File: python/lsst/dax/imgserv/api_soda.py (hunk 3 above)
Comment (kennywlo) [reply python/lsst/dax/imgserv/api_soda.py]: Agree. Let me change the prefix to `img` instead.  Started out with naming as suggested by blocked by 'async' is a reserved keyword.

File: python/lsst/dax/imgserv/templates/vosi_capabilities.xml (hunk 9 above)
Comment (kennywlo) [reply python/lsst/dax/imgserv/templates/vosi_capabilities.xml]: ImgServ or another closely VO service (like metaserv) will have to provide SIAv2 support; in fact, it's probably first thing users want to use, as part of prep before invoking SODA operations.  DM-14330 been opened for this work, that I will get to in near future.

File: python/lsst/dax/imgserv/vo/imageSODA.py (hunk 10 above)
Comment (kennywlo) [reply python/lsst/dax/imgserv/vo/imageSODA.py]: See above comment.

File: tests/engineFactory.py (hunk 12 above)
Comment (kennywlo) [reply tests/engineFactory.py]: These are the dependencies of old unit tests that use sqlachemy and mysql connect to metaserv to unit test FITS files and its metadata (headers).  As @brianv0 pointed out, that's some really OLD proprietary tests, that are no longer relevant in our new VO way of doing things.  The metadata tests, for instance, will be replaced by unit tests with sia.

File: tests/conftest.py (hunk 11 above)
Comment (kennywlo) [reply tests/conftest.py]: Yeah.

File: tests/engineFactory.py (hunk 12 above)
Comment (timj) [reply tests/engineFactory.py]: Deleting these sounds like it should be a distinct commit with an explanation as to why they are being deleted in the commit message.

**Comment by kennywlo** [comment]
Ok, got the message on commits.  After addressing other input, I will do the Git work to separate out the commits as suggested: 1) removal of old unit tests 2) sodalint addition 3) fetching host name in full soda URLs (...etc).

File: python/lsst/dax/imgserv/api_soda.py (hunk 4 above)
Comment (kennywlo) [reply python/lsst/dax/imgserv/api_soda.py]: yep, changing to use url_for() in flask as for /capabilities and /examples, greatly simplifying the logic here.

File: python/lsst/dax/imgserv/api_soda.py (hunk 5 above)
Comment (kennywlo) [reply python/lsst/dax/imgserv/api_soda.py]: Removed.  Was useful earlier during coding.

File: python/lsst/dax/imgserv/templates/dali_examples.xml (hunk 8 above)
Comment (kennywlo) [reply python/lsst/dax/imgserv/templates/dali_examples.xml]: The standard calling sequence for user is home portal for imgserv, seeing it supports VO, then /availability, followed by /capabilities, then /examples [optional, not widely supported yet, including sodalint].

File: tests/engineFactory.py (hunk 12 above)
Comment (kennywlo) [reply tests/engineFactory.py]: agree.

File: tests/sodalint/LICENSE (hunk 1 above)
Comment (kennywlo) [reply tests/sodalint/LICENSE]: agree.

File: python/lsst/dax/imgserv/api_soda.py (hunk 6 above)
Comment (kennywlo) [reply python/lsst/dax/imgserv/api_soda.py]: I was thinking the same earlier, until I checked out the reference SODA service (https://casda.csiro.au/casda_data_access/data/) in sodalint is returning, which is http 200 for /sync and /async (blank requests) along with the service descriptor.

**Comment by kennywlo** [comment]
I've done all the changes as advised, including separating the commits.
